File: src/MyServiceBus/Serialization/RawJsonMessageSerializer.cs
using System.Text.Json;
using System.Text.Json.Serialization.Metadata;

namespace MyServiceBus.Serialization;

public class RawJsonMessageSerializer : IMessageSerializer, IMessageSerializerMetadata
{
    private readonly IMessageHeaderConvention _headerConvention;
    private readonly JsonSerializerOptions _jsonSerializerOptions;

    public RawJsonMessageSerializer()
        : this(JsonSerializationDefaults.CreateOptions(), MassTransitHeaderConvention.Instance)
    {
    }

    public RawJsonMessageSerializer(IMessageHeaderConvention headerConvention)
        : this(JsonSerializationDefaults.CreateOptions(), headerConvention)
    {
    }

    public RawJsonMessageSerializer(
        JsonSerializerOptions jsonSerializerOptions,
        IMessageHeaderConvention? headerConvention = null)
    {
        ArgumentNullException.ThrowIfNull(jsonSerializerOptions);
        _jsonSerializerOptions = jsonSerializerOptions;
        _headerConvention = headerConvention ?? MassTransitHeaderConvention.Instance;
    }

    public string ContentType => InboundMessageResolver.RawJsonContentType;

    public MessageEnvelopeMode EnvelopeMode => MessageEnvelopeMode.Raw;

    public MessageBody GetMessageBody<T>(MessageSerializationContext<T> context) where T : class
    {
        context.Headers[_headerConvention.ContentTypeHeader] = ContentType;
        var typeInfo = _jsonSerializerOptions.GetTypeInfo(typeof(T)) as JsonTypeInfo<T>
            ?? throw new InvalidOperationException($"JSON metadata is not configured for {typeof(T)}.");
        return new ByteArrayMessageBody(JsonSerializer.SerializeToUtf8Bytes(context.Message!, typeInfo));
    }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/serialization/RawJsonMessageSerializer.java
package com.myservicebus.serialization;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;

public class RawJsonMessageSerializer implements MessageSerializer, MessageSerializerMetadata {
    private final ObjectMapper mapper;
    private final MessageHeaderConvention headerConvention;

    @Override
    public String getContentType() {
        return DefaultInboundMessageResolver.RAW_JSON_CONTENT_TYPE;
    }

    @Override
    public MessageEnvelopeMode getEnvelopeMode() {
        return MessageEnvelopeMode.RAW;
    }

    public RawJsonMessageSerializer() {
        this(JsonSerializationDefaults.createObjectMapper(), MassTransitHeaderConvention.INSTANCE);
    }

    public RawJsonMessageSerializer(MessageHeaderConvention headerConvention) {
        this(JsonSerializationDefaults.createObjectMapper(), headerConvention);
    }

    public RawJsonMessageSerializer(ObjectMapper mapper) {
        this(mapper, MassTransitHeaderConvention.INSTANCE);
    }

    public RawJsonMessageSerializer(ObjectMapper mapper, MessageHeaderConvention headerConvention) {
        if (mapper == null) {
            throw new IllegalArgumentException("mapper must not be null");
        }
        this.mapper = mapper;
        this.headerConvention = headerConvention;
    }

    @Override
    public <T> MessageBody getMessageBody(MessageSerializationContext<T> context) throws IOException {
        context.getHeaders().put(headerConvention.getContentTypeHeader(), getContentType());
        return new ByteArrayMessageBody(mapper.writeValueAsBytes(context.getMessage()));
    }
}
